airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/state_converters/datetime_stream_state_converter.py

```diff
@@ -113,23 +137,36 @@ def convert_to_sequential_state(self, cursor_field: CursorField, stream_state: M
         """
         if self.is_state_message_compatible(stream_state):
             legacy_state = stream_state.get("legacy", {})
-            if slices := stream_state.pop("slices", None):
-                latest_complete_time = self._get_latest_complete_time(slices)
-                if latest_complete_time:
-                    legacy_state.update({cursor_field.cursor_field_key: self.output_format(latest_complete_time)})
+            latest_complete_time = self._get_latest_complete_time(stream_state["low_water_mark"], stream_state.get("slices", []))
+            if latest_complete_time is not None:
+                legacy_state.update({cursor_field.cursor_field_key: self.output_format(latest_complete_time)})
             return legacy_state or {}
         else:
             return stream_state
 
-    def _get_latest_complete_time(self, slices: List[MutableMapping[str, Any]]) -> Optional[datetime]:
+    def _get_latest_complete_time(self, previous_sync_end: datetime, slices: List[MutableMapping[str, Any]]) -> Optional[datetime]:
         """
         Get the latest time before which all records have been processed.
         """
         if slices:
-            first_interval = self.merge_intervals(slices)[0][self.END_KEY]
-            return first_interval
+            merged_intervals = self.merge_intervals(slices)
+            first_interval = merged_intervals[0]
+            if previous_sync_end < first_interval[self.START_KEY]:
```

Comment:
I would have assume that nothing would have changed except from https://github.com/airbytehq/airbyte/pull/34131/files#diff-44999ccc78e8a64a5a79f8352f8aa2a45d8dde6ecaa92d3dff0cc773de104197R118. How can this case be possible? It feels like since we add an interval with `low_water_mark ` as an upper boundary, we should never have this case. Do we fear that connectors might start syncing earlier slices?

Reply:
This is handling what I assume would be an unusual case, which is that the start date was changed to something more recent.